src/range.rs

```diff
@@ -0,0 +1,82 @@
+use core::ops::{ RangeBounds, Bound };
+use core::iter::FusedIterator;
+use crate::Itertools;
+
+/// An iterator over a range of values.
+///
+/// Acquired by the [`range`] function or the
+/// [`Itertools::range`] method.
+pub struct Range<I, R> {
+    range: R,
+    internal: I,
+    counter: usize,
+}
+
+impl<I, R> Clone for Range<I, R> 
+    where I: Clone, R: Clone 
+{
+    fn clone(&self) -> Self {
+        Range {
+            counter: self.counter,
+            range: self.range.clone(),
+            internal: self.internal.clone(),
+        }
+    }
+}
+
+impl<T, I, R> Iterator for Range<I, R>
+    where I: Iterator<Item = T>, 
+          R: RangeBounds<usize> 
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.counter == 0 {
```

Comment:
I am thinking if we could avoid `self.counter` entirely, and just store and modify the bounds accordingly... Not sure if this is a good idea, though.